File: vBox/GoogleMapsViewController.m
//
//  ViewController.m
//  vBox
//
//  Created by Rosbel Sanroman on 10/2/14.
//  Copyright (c) 2014 rosbelSanroman. All rights reserved.
//

#import "GoogleMapsViewController.h"
#import <GoogleMaps/GoogleMaps.h>
#import "OLGhostAlertView.h"
#import "SVProgressHUD.h"
#import "MyStyleKit.h"

@interface GoogleMapsViewController ()

@property (weak, nonatomic) IBOutlet UILabel *bluetoothRequiredLabel;
@property (weak, nonatomic) IBOutlet UIView *infoView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *mapViewToInfoViewConstraint;

@end

@implementation GoogleMapsViewController{
	GMSCameraPosition *camera;
	GMSMutablePath *completePath;
	GMSPolyline* polyline;
	NSMutableArray *pastLocations;
	BOOL followMe;
	AppDelegate *appDelegate;
	NSManagedObjectContext *context;
	Trip *currentTrip;
	double sumSpeed;
	double maxSpeed;
	double minSpeed;
	NSArray *styles;
	CGRect infoViewFrame;
	CGRect mapViewFrame;
	CGRect infoViewHiddenOffScreen;
	BOOL showSpeed;
	BOOL bleOn;
}

#pragma mark - UIView Delegate Methods

- (void)viewDidLoad {
	[super viewDidLoad];
	
	appDelegate = [[UIApplication sharedApplication] delegate];
	context = [appDelegate managedObjectContext];
	
	currentTrip = [NSEntityDescription insertNewObjectForEntityForName:@"Trip" inManagedObjectContext:context];
	[currentTrip setStartTime:[NSDate date]];
	
	completePath = [GMSMutablePath path];
	pastLocations = [NSMutableArray array];
	
	styles = @[[GMSStrokeStyle solidColor:[UIColor colorWithRed:0.2666666667 green:0.4666666667 blue:0.6 alpha:1]],[GMSStrokeStyle solidColor:[UIColor colorWithRed:0.6666666667 green:0.8 blue:0.8 alpha:1]]];
	
	sumSpeed = 0;
	maxSpeed = 0;
	minSpeed = DBL_MAX;
	followMe = YES;
	showSpeed = YES;
	bleOn = NO;
	
	self.speedOrDistanceLabel.userInteractionEnabled = YES;
	UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(speedLabelTapped)];
	[self.speedOrDistanceLabel addGestureRecognizer:tapGesture];
	
	self.bluetoothDiagnostics = [NSMutableDictionary dictionary];
	
	[self setUpLocationManager];
	
	[self setUpGoogleMaps];
	
	[self setUpUIButtons];
}

-(void)viewDidLayoutSubviews
{
	infoViewFrame = self.infoView.frame;
	mapViewFrame = self.MapView.frame;
	infoViewHiddenOffScreen = self.infoView.frame;
	infoViewHiddenOffScreen.origin.y = [[UIScreen mainScreen] bounds].size.height;
	
	[self updateViewsBasedOnBLEButtonState:bleOn animate:NO];
}

-(void)viewWillDisappear:(BOOL)animated
{
	[_MapView clear];
	_MapView = nil;
	[_locationManager stopUpdatingLocation];
	
	[self cleanUpBluetoothManager];
	
	//Delete If no locations were recorded
	if(currentTrip.gpsLocations.count == 0)
	{
		[[appDelegate managedObjectContext] deleteObject:currentTrip];
		[appDelegate saveContext];
		return;
	}
	[currentTrip setEndTime:[NSDate date]];
	unsigned long count = currentTrip.gpsLocations.count;
	double avgSpeed = count > 0 ? sumSpeed / count : 0;
	[currentTrip setAvgSpeed:[NSNumber numberWithDouble:avgSpeed]];
	[currentTrip setMaxSpeed:[NSNumber numberWithDouble:maxSpeed]];
	[currentTrip setMinSpeed:[NSNumber numberWithDouble:minSpeed]];
	[currentTrip setTotalMiles:[NSNumber numberWithDouble:GMSGeometryLength(completePath)*0.000621371]];
	[[appDelegate drivingHistory] addTripsObject:currentTrip];
	[appDelegate saveContext];
	
	[super viewWillDisappear:animated];
}

#pragma mark - SetUp Methods

-(void)setUpBluetoothManager
{
	self.bluetoothManager = [[BLEManager alloc] init];
	self.bluetoothManager.delegate = self;
}

-(void)setUpLocationManager
{
	_locationManager = [[CLLocationManager alloc] init];
	
	[_locationManager setDelegate:self];
	_locationManager.distanceFilter = kCLDistanceFilterNone; //Best Accuracy
	_locationManager.desiredAccuracy = kCLLocationAccuracyBestForNavigation;
	_locationManager.activityType = CLActivityTypeAutomotiveNavigation;
	_locationManager.pausesLocationUpdatesAutomatically = YES;//help save battery life when user is stopped
	
	[_locationManager requestWhenInUseAuthorization];
	[_locationManager requestAlwaysAuthorization];
	[_locationManager startUpdatingLocation];
}

-(void)setUpGoogleMaps
{
	camera = [GMSCameraPosition cameraWithLatitude:39.490179
										 longitude:-98.081992
											  zoom:3];
	
	[_MapView setPadding:UIEdgeInsetsMake(85, 0, 0, 5)];
	[_MapView setCamera:camera];
	_MapView.myLocationEnabled = YES;
	_MapView.settings.myLocationButton = YES;
	_MapView.settings.compassButton = YES;
	[_MapView setDelegate:self];
	
	polyline = [GMSPolyline polylineWithPath:completePath];
	polyline.strokeColor = [UIColor grayColor];
	polyline.strokeWidth = 5.0;
	polyline.geodesic = YES;
	polyline.map = self.MapView;
}

-(void)setUpUIButtons
{
	self.bleButton.layer.masksToBounds = YES;
	self.bleButton.layer.cornerRadius = 5.0;
	
	self.stopRecordingButton.layer.masksToBounds = YES;
	self.stopRecordingButton.layer.cornerRadius = 5.0;
	
	[self.stopRecordingButton setBackgroundImage:[MyStyleKit imageOfVBoxButtonWithButtonColor:self.stopRecordingButton.backgroundColor] forState:UIControlStateNormal];
	
	self.speedOrDistanceLabel.layer.masksToBounds = YES;
	self.speedOrDistanceLabel.layer.cornerRadius = 5.0;
}

#pragma mark - Google Maps View Delegate

-(void)mapView:(GMSMapView *)mapView willMove:(BOOL)gesture
{
	if(gesture)
	{
		followMe = NO;
	}
}

-(BOOL)didTapMyLocationButtonForMapView:(GMSMapView *)mapView
{
	followMe = YES;
	return NO;
}


#pragma mark - Button Action Methods

- (IBAction)stopRecordingButtonTapped:(id)sender {
	
	[self.navigationController popViewControllerAnimated:YES];
	[self.delegate didTapStopRecordingButton];
}

- (IBAction)bleButtonTapped:(UIButton *)sender {
	bleOn = !bleOn;
	if(bleOn)
	{
		[sender setImage:[UIImage imageNamed:@"bleOn"] forState:UIControlStateNormal];
		[self setUpBluetoothManager];
	}else
	{
		[sender setImage:[UIImage imageNamed:@"bleOff"] forState:UIControlStateNormal];
		[self cleanUpBluetoothManager];
	}
	[self updateViewsBasedOnBLEButtonState:bleOn animate:YES];
}

#pragma mark - Speed Label Methods

-(void)speedLabelTapped
{
	showSpeed = !showSpeed;
	[self updateSpeedLabelWithLocation:[pastLocations lastObject]];
}

-(void)updateSpeedLabelWithLocation:(CLLocation *)lastLocation
{
	if(showSpeed)
	{
		self.speedOrDistanceLabel.text = [NSString stringWithFormat:@" %.2f mph",lastLocation.speed * 2.23694];
	}else
	{
		self.speedOrDistanceLabel.text = [NSString stringWithFormat:@" %.2f mi",GMSGeometryLength(completePath) * 0.000621371];
	}
}

#pragma mark - CLLocation Delegate

-(void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations
{
	
	unsigned long objCount = [locations count];
	unsigned long prevCount = [pastLocations count];
	
	CLLocation *newestLocation = [locations lastObject];
	CLLocation *prevLocation;

	if(followMe && prevCount < 1 && newestLocation.horizontalAccuracy < 70)
	{
		[self.MapView animateToLocation:newestLocation.coordinate];
		if(self.MapView.camera.zoom < 10)
			[self.MapView animateToZoom:15];
	}
	//Ignore bad Accuracy
	if(newestLocation.horizontalAccuracy > 30) //maybe give user tolerance for bad accuracy?
	{
		return;
	}
	
	if(objCount > 1)
	{
		prevLocation = [locations objectAtIndex:objCount - 2];
	}else if (prevCount > 1)
	{
		prevLocation = [pastLocations objectAtIndex:prevCount -2];
	}else
	{
		prevLocation = newestLocation;
	}
	
	//Update speed Label
	double speedMPH = ([newestLocation speed] * 2.236936284);
	speedMPH = speedMPH >= 0 ? speedMPH : 0;
	
	if(speedMPH < minSpeed)
	{
		minSpeed = speedMPH;
	}
	if(speedMPH > maxSpeed)
	{
		maxSpeed = speedMPH;
	}
	sumSpeed += speedMPH;
	
	
	[completePath addCoordinate:newestLocation.coordinate];
	
	[self updateSpeedLabelWithLocation:newestLocation];
	
	[self logLocation:newestLocation persistent:YES];
	
	[polyline setPath:completePath];
	
	double tolerance = powf(10.0,(-0.301*self.MapView.camera.zoom)+9.0731) / 2500.0;
	NSArray *lengths = @[@(tolerance),@(tolerance*1.5)];
	polyline.spans = GMSStyleSpans(polyline.path, styles, lengths, kGMSLengthGeodesic);
	
	if(followMe)
	{
		[_MapView animateToLocation:newestLocation.coordinate];
	}
}


-(void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
{
	if([error domain] == kCLErrorDomain )
	{
		return; //sometimes I get this error // fix later?
	}
	UIAlertView *errorAlert = [[UIAlertView alloc]initWithTitle:[error localizedDescription] message:@"There was an error retrieving your location" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles: nil];
	[errorAlert show];
}

- (void)didReceiveMemoryWarning {
	[super didReceiveMemoryWarning];
	// Dispose of any resources that can be recreated.
}

#pragma mark - UICollection Data Source Delegate

-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
	return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
	return self.bluetoothDiagnostics.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
	NSArray *keys = [[self.bluetoothDiagnostics allKeys] sortedArrayUsingSelector:@selector(compare:)];
	NSString *key = [keys objectAtIndex:indexPath.row];
	
	UICollectionViewCell *cell;
	
	cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"collectionCell" forIndexPath:indexPath];
	UILabel *keyLabel = (UILabel *)[cell viewWithTag:1];
	UILabel *valLabel = (UILabel *)[cell viewWithTag:2];
	keyLabel.text = key;
	valLabel.text = [NSString stringWithFormat:@"%@",(NSNumber *)[self.bluetoothDiagnostics objectForKey:key]];
	return cell;
}

#pragma mark - BLEManager Delegate

-(void)didChangeBluetoothState:(BLEState)state
{
	switch(state)
	{
		case BLEStateOn:
			self.bluetoothRequiredLabel.hidden = YES;
			if([[NSUserDefaults standardUserDefaults] boolForKey:@"connectToOBD"])
			{
				[self.bluetoothManager scanForPeripheralType:PeripheralTypeOBDAdapter];
			}
			else
			{
				[self.bluetoothManager scanForPeripheralType:PeripheralTypeBeagleBone];
			}
			break;
		case BLEStateOff:
			[SVProgressHUD showErrorWithStatus:@"Bluetooth Off"];
			break;
		case BLEStateResetting:
			[SVProgressHUD showErrorWithStatus:@"Bluetooth Resetting"];
			break;
		case BLEStateUnauthorized:
			[SVProgressHUD showErrorWithStatus:@"Bluetooth Unauthorized"];
			break;
		case BLEStateUnkown:
			[SVProgressHUD showErrorWithStatus:@"Bluetooth State Uknown"];
			break;
		case BLEStateUnsupported:
			[SVProgressHUD showErrorWithStatus:@"Bluetooth Unsupported"];
			break;
	}
	
	if(state != BLEStateOn)
	{
		self.bluetoothRequiredLabel.hidden = NO;
	}
}

-(void)didConnectPeripheral
{
	[SVProgressHUD showSuccessWithStatus:@"Connected"];
}

-(void)didBeginScanningForPeripheral
{
	[SVProgressHUD showWithStatus:@"Scanning.."];
}

-(void)didDisconnectPeripheral
{
	[SVProgressHUD showErrorWithStatus:@"Disconnected"];
	//try reconnecting
}

-(void)didStopScanning
{
	
}

-(void)didUpdateDiagnosticForKey:(NSString *)key withValue:(NSNumber *)value
{
	[self.bluetoothDiagnostics setObject:value forKey:key];
	[self.collectionView reloadData];
}

-(void)didUpdateDiagnosticForKey:(NSString *)key withMultipleValues:(NSArray *)values
{
	
}

#pragma mark - Layout Methods

-(void)updateViewsBasedOnBLEButtonState:(BOOL) state animate:(BOOL)animate
{
	if(state)
	{
		if(animate)
		{
			[UIView beginAnimations:@"ShowInfoView" context:nil];
			[UIView setAnimationBeginsFromCurrentState:NO];
			[UIView setAnimationCurve:UIViewAnimationCurveEaseOut];
			[UIView setAnimationDelegate:self];
			[UIView setAnimationDuration:0.25];
			[UIView setAnimationBeginsFromCurrentState:YES];
			[UIView setAnimationWillStartSelector:@selector(animationDidStart:context:)];
			[UIView setAnimationDidStopSelector:@selector(animationDidStop:finished:context:)];
		}
		[self.infoView setFrame:infoViewFrame];
		[self.MapView setFrame:mapViewFrame];
		if(animate)
		{
			[UIView commitAnimations];
		}
	}else
	{
		if(animate)
		{
			[UIView beginAnimations:@"HideInfoView" context:nil];
			[UIView setAnimationCurve:UIViewAnimationCurveEaseOut];
			[UIView setAnimationDuration:0.25];
			[UIView setAnimationDelegate:self];
			[UIView setAnimationWillStartSelector:@selector(animationDidStart:context:)];
			[UIView setAnimationDidStopSelector:@selector(animationDidStop:finished:context:)];
		}
		[self.infoView setFrame:infoViewHiddenOffScreen];
		[self.MapView setFrame:[UIScreen mainScreen].bounds];
		if(animate)
		{
			[UIView commitAnimations];
		}
	}
}
- (void)animationDidStart:(NSString *)animationID context:(void *)context
{
	if([animationID isEqualToString:@"ShowInfoView"])
	{
		[self.infoView setHidden:NO];
	}
	self.bleButton.enabled = NO;
}

- (void)animationDidStop:(NSString *)animationID finished:(NSNumber *)finished context:(void *)context
{
	if([animationID isEqualToString:@"HideInfoView"])
	{
		[self.infoView setHidden:YES];
	}
	self.bleButton.enabled = YES;
}

#pragma mark - Core Data

-(void)logLocation:(CLLocation *)location persistent:(Boolean)persist
{
	double lat = location.coordinate.latitude;
	double lng = location.coordinate.longitude;
	double speedMPH = location.speed >= 0 ? location.speed * 2.236936284 : 0;
	NSDate *time = location.timestamp;
	
	if(persist)
	{
		GPSLocation *newLocation = [NSEntityDescription insertNewObjectForEntityForName:@"GPSLocation" inManagedObjectContext:context];
		[newLocation setLatitude:[NSNumber numberWithDouble:lat]];
		[newLocation setLongitude:[NSNumber numberWithDouble:lng]];
		[newLocation setSpeed:[NSNumber numberWithDouble:speedMPH]];
		[newLocation setMetersFromStart:[NSNumber numberWithDouble:GMSGeometryLength(completePath)]];
		[newLocation setTimestamp:time];
		[newLocation setTripInfo:currentTrip];
		
		if(self.bluetoothManager.connected)
		{
			
			BluetoothData *bleData = [NSEntityDescription insertNewObjectForEntityForName:@"BluetoothData" inManagedObjectContext:context];
			NSNumber *bleSpeedMPH =[self.bluetoothDiagnostics objectForKey:@"Speed"];
			bleSpeedMPH = bleSpeedMPH ? [NSNumber numberWithDouble:(bleSpeedMPH.doubleValue * 0.621371)] : bleSpeedMPH;
			[bleData setSpeed:bleSpeedMPH];
			[bleData setAmbientTemp:[self.bluetoothDiagnostics objectForKey:@"Ambient Temp"]];
			[bleData setBarometric:[self.bluetoothDiagnostics objectForKey:@"Barometric"]];
			[bleData setRpm:[self.bluetoothDiagnostics objectForKey:@"RPM"]];
			[bleData setIntakeTemp:[self.bluetoothDiagnostics objectForKey:@"Intake Temp"]];
			[bleData setFuel:[self.bluetoothDiagnostics objectForKey:@"Fuel"]];
			[bleData setEngineLoad:[self.bluetoothDiagnostics objectForKey:@"Engine Load"]];
			[bleData setDistance:[self.bluetoothDiagnostics objectForKey:@"Distance"]];
			[bleData setCoolantTemp:[self.bluetoothDiagnostics objectForKey:@"Coolant Temp"]];
			[bleData setThrottle:[self.bluetoothDiagnostics objectForKey:@"Throttle"]];
			//Set AccelX,Y,Z
			[newLocation setBluetoothInfo:bleData];
		}
		[appDelegate saveContext];
		//Check if Bluetooth is on, and populate DB with recent values
	}
	
	[pastLocations addObject:location];
}


#pragma mark - Clean UP

-(void)cleanUpBluetoothManager
{
	
	if(self.bluetoothManager.connected)
	{
		[self.bluetoothManager disconnect];
	}
	
	[self.bluetoothManager stopScanning];
	[self.bluetoothManager stopAdvertisingPeripheral];
	self.bluetoothManager = nil;
	
	[SVProgressHUD dismiss];
}


@end

File: vBox/ViewControllers/GoogleMapsViewController.swift
//
//  GoogleMapsViewController.swift
//  vBox
//
//  Swift implementation of main driving/recording view
//

import UIKit
import CoreLocation
import GoogleMaps
import Combine

// MARK: - Delegate Protocol

@objc protocol GoogleMapsViewControllerSwiftDelegate: AnyObject {
    func didTapStopRecordingButton()
}

// MARK: - Google Maps View Controller

final class GoogleMapsViewControllerSwift: UIViewController {

    // MARK: - IBOutlets

    @IBOutlet private weak var MapView: GMSMapView!
    @IBOutlet private weak var stopRecordingButton: UIButton!
    @IBOutlet private weak var speedOrDistanceLabel: UILabel!
    @IBOutlet private weak var collectionView: UICollectionView!
    @IBOutlet private weak var bleButton: UIButton!
    @IBOutlet private weak var bluetoothRequiredLabel: UILabel!
    @IBOutlet private weak var infoView: UIView!
    @IBOutlet private weak var mapViewToInfoViewConstraint: NSLayoutConstraint!

    // MARK: - Properties

    @objc weak var delegate: GoogleMapsViewControllerSwiftDelegate?

    private var locationManager: CLLocationManager!
    private var bleManager: BLEManagerSwift?
    private var cancellables = Set<AnyCancellable>()

    private var completePath: GMSMutablePath!
    private var polyline: GMSPolyline!
    private var prevLocation: CLLocation?
    private var isFollowingMe = true
    private var showSpeed = true
    private var isBLEOn = false

    private var currentTrip: Trip!
    private var sumSpeed: Double = 0
    private var maxSpeed: Double = 0
    private var minSpeed: Double = .greatestFiniteMagnitude

    private var infoViewFrame: CGRect = .zero
    private var mapViewFrame: CGRect = .zero
    private var infoViewHiddenOffScreen: CGRect = .zero

    private var bluetoothDiagnostics: [String: NSNumber] = [:]

    private var appDelegate: AppDelegate {
        return UIApplication.shared.delegate as! AppDelegate
    }

    private var context: NSManagedObjectContext {
        return appDelegate.managedObjectContext
    }

    private let pathStyles: [GMSStrokeStyle] = [
        .solidColor(UIColor(red: 0.2666666667, green: 0.4666666667, blue: 0.6, alpha: 1)),
        .solidColor(UIColor(red: 0.6666666667, green: 0.8, blue: 0.8, alpha: 1))
    ]

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        // Create new trip
        currentTrip = NSEntityDescription.insertNewObject(forEntityName: "Trip", into: context) as? Trip
        currentTrip.startTime = Date()

        completePath = GMSMutablePath()

        setupUI()
        setupLocationManager()
        setupGoogleMaps()
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        infoViewFrame = infoView.frame
        mapViewFrame = MapView.frame
        infoViewHiddenOffScreen = infoView.frame
        infoViewHiddenOffScreen.origin.y = UIScreen.main.bounds.height

        updateViewsBasedOnBLEState(isBLEOn, animate: false)
    }

    override func viewWillDisappear(_ animated: Bool) {
        MapView.clear()
        locationManager.stopUpdatingLocation()

        cleanUpBluetoothManager()

        // Delete trip if no locations recorded
        if currentTrip.gpsLocations?.count == 0 {
            context.delete(currentTrip)
            appDelegate.saveContext()
            super.viewWillDisappear(animated)
            return
        }

        // Save trip statistics
        currentTrip.endTime = Date()
        let count = currentTrip.gpsLocations?.count ?? 0
        let avgSpeed = count > 0 ? sumSpeed / Double(count) : 0
        currentTrip.avgSpeed = NSNumber(value: avgSpeed)
        currentTrip.maxSpeed = NSNumber(value: maxSpeed)
        currentTrip.minSpeed = NSNumber(value: minSpeed)
        currentTrip.totalMiles = NSNumber(value: GMSGeometryLength(completePath) * 0.000621371)

        appDelegate.drivingHistory.addToTrips(currentTrip)
        appDelegate.saveContext()

        super.viewWillDisappear(animated)
    }

    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .default
    }

    // MARK: - Setup

    private func setupUI() {
        // Speed label tap gesture
        speedOrDistanceLabel.isUserInteractionEnabled = true
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(speedLabelTapped))
        speedOrDistanceLabel.addGestureRecognizer(tapGesture)

        // Button styling
        bleButton.layer.masksToBounds = true
        bleButton.layer.cornerRadius = 5.0

        stopRecordingButton.layer.masksToBounds = true
        stopRecordingButton.layer.cornerRadius = 5.0
        let buttonColor = UIColor(red: 0, green: 122.0 / 255.0, blue: 1, alpha: 1)
        stopRecordingButton.setBackgroundImage(MyStyleKit.image(ofVBoxButtonWithButtonColor: buttonColor), for: .normal)

        speedOrDistanceLabel.layer.masksToBounds = true
        speedOrDistanceLabel.layer.cornerRadius = 5.0
    }

    private func setupLocationManager() {
        locationManager = CLLocationManager()
        locationManager.delegate = self
        locationManager.distanceFilter = kCLDistanceFilterNone
        locationManager.desiredAccuracy = kCLLocationAccuracyBestForNavigation
        locationManager.activityType = .automotiveNavigation
        locationManager.pausesLocationUpdatesAutomatically = true

        locationManager.requestWhenInUseAuthorization()
        locationManager.requestAlwaysAuthorization()
        locationManager.startUpdatingLocation()
    }

    private func setupGoogleMaps() {
        let camera = GMSCameraPosition(latitude: 39.490179, longitude: -98.081992, zoom: 3)

        MapView.padding = UIEdgeInsets(top: 85, left: 0, bottom: 0, right: 5)
        MapView.camera = camera
        MapView.isMyLocationEnabled = true
        MapView.settings.myLocationButton = true
        MapView.settings.compassButton = true
        MapView.delegate = self

        polyline = GMSPolyline(path: completePath)
        polyline.strokeColor = .gray
        polyline.strokeWidth = 5.0
        polyline.geodesic = true
        polyline.map = MapView
    }

    private func setupBluetoothManager() {
        bleManager = BLEManagerSwift()

        // Subscribe to state changes
        bleManager?.statePublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.handleBLEStateChange(state)
            }
            .store(in: &cancellables)

        // Subscribe to connection changes
        bleManager?.connectionPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] isConnected in
                if isConnected {
                    SVProgressHUD.showSuccess(withStatus: "Connected")
                } else {
                    SVProgressHUD.showError(withStatus: "Disconnected")
                }
            }
            .store(in: &cancellables)

        // Subscribe to diagnostics
        bleManager?.diagnosticsPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] diagnostics in
                self?.updateBluetoothDiagnostics(diagnostics)
            }
            .store(in: &cancellables)

        // Subscribe to scanning status
        bleManager?.$isScanning
            .receive(on: DispatchQueue.main)
            .sink { isScanning in
                if isScanning {
                    SVProgressHUD.show(withStatus: "Scanning...")
                }
            }
            .store(in: &cancellables)
    }

    // MARK: - BLE State Handling

    private func handleBLEStateChange(_ state: BLEState) {
        switch state {
        case .on:
            bluetoothRequiredLabel.isHidden = true
            let peripheralType: PeripheralType = UserDefaults.standard.bool(forKey: "connectToOBD") ? .obdAdapter : .beagleBone
            _ = bleManager?.scan(for: peripheralType)
        case .off:
            SVProgressHUD.showError(withStatus: "Bluetooth Off")
            bluetoothRequiredLabel.isHidden = false
        case .resetting:
            SVProgressHUD.showError(withStatus: "Bluetooth Resetting")
            bluetoothRequiredLabel.isHidden = false
        case .unauthorized:
            SVProgressHUD.showError(withStatus: "Bluetooth Unauthorized")
            bluetoothRequiredLabel.isHidden = false
        case .unknown:
            SVProgressHUD.showError(withStatus: "Bluetooth State Unknown")
            bluetoothRequiredLabel.isHidden = false
        case .unsupported:
            SVProgressHUD.showError(withStatus: "Bluetooth Unsupported")
            bluetoothRequiredLabel.isHidden = false
        }
    }

    private func updateBluetoothDiagnostics(_ diagnostics: VehicleDiagnostics) {
        if let speed = diagnostics.speed {
            bluetoothDiagnostics["Speed"] = NSNumber(value: speed)
        }
        if let rpm = diagnostics.rpm {
            bluetoothDiagnostics["RPM"] = NSNumber(value: rpm)
        }
        if let fuel = diagnostics.fuelLevel {
            bluetoothDiagnostics["Fuel"] = NSNumber(value: fuel)
        }
        if let coolant = diagnostics.coolantTemp {
            bluetoothDiagnostics["Coolant Temp"] = NSNumber(value: coolant)
        }
        if let intake = diagnostics.intakeTemp {
            bluetoothDiagnostics["Intake Temp"] = NSNumber(value: intake)
        }
        if let ambient = diagnostics.ambientTemp {
            bluetoothDiagnostics["Ambient Temp"] = NSNumber(value: ambient)
        }
        if let load = diagnostics.engineLoad {
            bluetoothDiagnostics["Engine Load"] = NSNumber(value: load)
        }
        if let throttle = diagnostics.throttlePosition {
            bluetoothDiagnostics["Throttle"] = NSNumber(value: throttle)
        }
        if let barometric = diagnostics.barometricPressure {
            bluetoothDiagnostics["Barometric"] = NSNumber(value: barometric)
        }
        if let distance = diagnostics.distanceSinceCodesCleared {
            bluetoothDiagnostics["Distance"] = NSNumber(value: distance)
        }

        collectionView.reloadData()
    }

    // MARK: - Actions

    @IBAction private func stopRecordingButtonTapped(_ sender: Any) {
        navigationController?.popViewController(animated: true)
        delegate?.didTapStopRecordingButton()
    }

    @IBAction private func bleButtonTapped(_ sender: UIButton) {
        isBLEOn.toggle()

        if isBLEOn {
            sender.setImage(UIImage(named: "bleOn"), for: .normal)
            setupBluetoothManager()
        } else {
            sender.setImage(UIImage(named: "bleOff"), for: .normal)
            cleanUpBluetoothManager()
        }

        updateViewsBasedOnBLEState(isBLEOn, animate: true)
    }

    @objc private func speedLabelTapped() {
        showSpeed.toggle()
        updateSpeedLabel(with: prevLocation)
    }

    // MARK: - UI Updates

    private func updateSpeedLabel(with location: CLLocation?) {
        guard let location = location else { return }

        if showSpeed {
            speedOrDistanceLabel.text = String(format: " %.2f mph", location.speed * 2.23694)
        } else {
            speedOrDistanceLabel.text = String(format: " %.2f mi", GMSGeometryLength(completePath) * 0.000621371)
        }
    }

    private func updateViewsBasedOnBLEState(_ state: Bool, animate: Bool) {
        let animations: () -> Void = {
            if state {
                self.infoView.frame = self.infoViewFrame
                self.MapView.frame = self.mapViewFrame
                self.infoView.isHidden = false
            } else {
                self.infoView.frame = self.infoViewHiddenOffScreen
                self.MapView.frame = UIScreen.main.bounds
            }
        }

        let completion: (Bool) -> Void = { _ in
            if !state {
                self.infoView.isHidden = true
            }
            self.bleButton.isEnabled = true
        }

        if animate {
            bleButton.isEnabled = false
            UIView.animate(withDuration: 0.25, delay: 0, options: .curveEaseOut, animations: animations, completion: completion)
        } else {
            animations()
            if !state {
                infoView.isHidden = true
            }
        }
    }

    // MARK: - Core Data

    private func logLocation(_ location: CLLocation, persistent: Bool) {
        let lat = location.coordinate.latitude
        let lon = location.coordinate.longitude
        let speedMPH = max(0, location.speed * 2.236936284)
        let altitude = location.altitude * 3.28084

        guard persistent else { return }

        let newLocation = NSEntityDescription.insertNewObject(forEntityName: "GPSLocation", into: context) as! GPSLocation
        newLocation.latitude = NSNumber(value: lat)
        newLocation.longitude = NSNumber(value: lon)
        newLocation.speed = NSNumber(value: speedMPH)
        newLocation.metersFromStart = NSNumber(value: GMSGeometryLength(completePath))
        newLocation.timestamp = location.timestamp
        newLocation.tripInfo = currentTrip
        newLocation.altitude = NSNumber(value: altitude)

        // Add Bluetooth data if connected
        if let bleManager = bleManager, bleManager.isConnected {
            let bleData = NSEntityDescription.insertNewObject(forEntityName: "BluetoothData", into: context) as! BluetoothData

            // Convert speed from km/h to mph
            if let speedKmh = bluetoothDiagnostics["Speed"] {
                bleData.speed = NSNumber(value: speedKmh.doubleValue * 0.621371)
            }

            bleData.ambientTemp = bluetoothDiagnostics["Ambient Temp"]
            bleData.barometric = bluetoothDiagnostics["Barometric"]
            bleData.rpm = bluetoothDiagnostics["RPM"]
            bleData.intakeTemp = bluetoothDiagnostics["Intake Temp"]
            bleData.fuel = bluetoothDiagnostics["Fuel"]
            bleData.engineLoad = bluetoothDiagnostics["Engine Load"]
            bleData.distance = bluetoothDiagnostics["Distance"]
            bleData.coolantTemp = bluetoothDiagnostics["Coolant Temp"]
            bleData.throttle = bluetoothDiagnostics["Throttle"]

            newLocation.bluetoothInfo = bleData
        }

        appDelegate.saveContext()
    }

    // MARK: - Cleanup

    private func cleanUpBluetoothManager() {
        if let bleManager = bleManager {
            if bleManager.isConnected {
                bleManager.disconnect()
            }
            bleManager.stopScanning()
        }

        cancellables.removeAll()
        bleManager = nil

        SVProgressHUD.dismiss()
    }
}

// MARK: - CLLocationManagerDelegate

extension GoogleMapsViewControllerSwift: CLLocationManagerDelegate {
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let newestLocation = locations.last else { return }

        // Initial camera positioning
        if isFollowingMe && prevLocation == nil && newestLocation.horizontalAccuracy < 70 {
            MapView.animate(toLocation: newestLocation.coordinate)
            if MapView.camera.zoom < 10 {
                MapView.animate(toZoom: 15)
            }
        }

        for location in locations {
            // Skip inaccurate readings
            guard location.horizontalAccuracy <= 30 else { continue }

            let speedMPH = max(0, newestLocation.speed * 2.236936284)

            // Update statistics
            if speedMPH < minSpeed {
                minSpeed = speedMPH
            }
            if speedMPH > maxSpeed {
                maxSpeed = speedMPH
            }
            sumSpeed += speedMPH

            updateSpeedLabel(with: newestLocation)
            logLocation(newestLocation, persistent: true)

            completePath.add(newestLocation.coordinate)
            polyline.path = completePath
        }

        // Update polyline styling
        let tolerance = pow(10.0, (-0.301 * Double(MapView.camera.zoom)) + 9.0731) / 2500.0
        let lengths = [NSNumber(value: tolerance), NSNumber(value: tolerance * 1.5)]
        polyline.spans = GMSStyleSpans(polyline.path!, pathStyles, lengths, .geodesic)

        prevLocation = newestLocation

        if isFollowingMe {
            MapView.animate(toLocation: newestLocation.coordinate)
        }
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        if (error as NSError).domain == kCLErrorDomain.description {
            return // Ignore CoreLocation domain errors
        }

        let alert = UIAlertController(
            title: error.localizedDescription,
            message: "There was an error retrieving your location",
            preferredStyle: .alert
        )
        alert.addAction(UIAlertAction(title: "OK", style: .default))
        present(alert, animated: true)
    }
}

// MARK: - GMSMapViewDelegate

extension GoogleMapsViewControllerSwift: GMSMapViewDelegate {
    func mapView(_ mapView: GMSMapView, willMove gesture: Bool) {
        if gesture {
            isFollowingMe = false
        }
    }

    func didTapMyLocationButton(for mapView: GMSMapView) -> Bool {
        isFollowingMe = true
        return false
    }
}

// MARK: - UICollectionViewDataSource

extension GoogleMapsViewControllerSwift: UICollectionViewDataSource {
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return bluetoothDiagnostics.count
    }

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let keys = bluetoothDiagnostics.keys.sorted()
        let key = keys[indexPath.row]

        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "collectionCell", for: indexPath)

        if let keyLabel = cell.viewWithTag(1) as? UILabel {
            keyLabel.text = key
        }
        if let valLabel = cell.viewWithTag(2) as? UILabel {
            valLabel.text = "\(bluetoothDiagnostics[key] ?? 0)"
        }

        return cell
    }
}

// MARK: - UICollectionViewDelegate

extension GoogleMapsViewControllerSwift: UICollectionViewDelegate {
    // Implement delegate methods if needed
}
